backend.native/tests/runtime/concurrent/shared0.kt

```diff
@@ -0,0 +1,301 @@
+    val semaphore: AtomicInt = AtomicInt(0)
+
+    val worker = Worker.start()
+
+    val future = doNotCollectInWorker(worker, semaphore)
+    while (semaphore.value < 1) {}
+    GC.collect()
+    semaphore.increment()
+
+    val value = future.result
+    assertEquals(3, value.get().a)
+    worker.requestTermination().result
+}
+
+class B1 {
+    lateinit var b2: SharedRef<B2>
+}
+
+data class B2(val b1: SharedRef<B1>)
+
+fun createCyclicGarbage(): Triple<AtomicReference<SharedRef<B1>?>, WeakReference<B1>, WeakReference<B2>> {
+    val b1 = SharedRef.create(B1())
+    val refB1: AtomicReference<SharedRef<B1>?> = AtomicReference(b1)
+    val weakB1 = WeakReference(b1.get())
+
+    val b2 = SharedRef.create(B2(b1))
+    val weakB2 = WeakReference(b2.get())
+
+    b1.get().b2 = b2
+
+    return Triple(refB1, weakB1, weakB2)
+}
+
+@Test fun doesNotCollectCyclicGarbage() {
+    val (refB1, weakB1, weakB2) = createCyclicGarbage()
+
+    refB1.value = null
+    GC.collect()
+
+    // If these asserts fail, that means SharedRef managed to clean up cyclic garbage all by itself.
+    assertNotNull(weakB1.get())
+    assertNotNull(weakB2.get())
+}
+
+fun createCrossThreadCyclicGarbage(
+    worker: Worker
+): Triple<AtomicReference<SharedRef<B1>?>, WeakReference<B1>, WeakReference<B2>> {
+    val b1 = SharedRef.create(B1())
+    val refB1: AtomicReference<SharedRef<B1>?> = AtomicReference(b1)
+    val weakB1 = WeakReference(b1.get())
+
+    val future = worker.execute(TransferMode.SAFE, { b1 }) { b1 ->
+        val b2 = SharedRef.create(B2(b1))
+        Pair(b2, WeakReference(b2.get()))
+    }
+    val (b2, weakB2) = future.result
+
+    b1.get().b2 = b2
+
+    return Triple(refB1, weakB1, weakB2)
+}
+
+@Test fun doesNotCollectCrossThreadCyclicGarbage() {
+    val worker = Worker.start()
+
+    val (refB1, weakB1, weakB2) = createCrossThreadCyclicGarbage(worker)
+
+    refB1.value = null
+    GC.collect()
+    worker.execute(TransferMode.SAFE, {}) { GC.collect() }.result
+
+    // If these asserts fail, that means SharedRef managed to clean up cyclic garbage all by itself.
+    assertNotNull(weakB1.get())
+    assertNotNull(weakB2.get())
+
+    worker.requestTermination().result
+}
+
+@Test fun concurrentAccess() {
+    val workerCount = 10
+    val workerUnlocker = AtomicInt(0)
+
+    val local = SharedRef.create(A(3))
+    assertEquals(3, local.get().a)
+
+    val workers = Array(workerCount) {
+        Worker.start()
+    }
+    val futures = Array(workers.size) {
+        workers[it].execute(TransferMode.SAFE, { Pair(local, workerUnlocker) }) { (local, workerUnlocker) ->
+            while (workerUnlocker.value < 1) {}
+
+            assertFailsWith<IncorrectDereferenceException> {
+                local.get()
+            }
+            Unit
+        }
+    }
+    workerUnlocker.increment()
+
+    for (future in futures) {
+        future.result
+    }
+
+    for (worker in workers) {
+        worker.requestTermination().result
+    }
+}
```

Comment:
No test for accessing `SharedRef` holding frozen object.
Object can be frozen both before and after `SharedRef` is created.